osu.Game/Online/Leaderboards/Leaderboard.cs

```diff
@@ -231,96 +145,183 @@ protected Leaderboard()
+                    case PlaceholderState.Unavailable:
+                        replacePlaceholder(new MessagePlaceholder(@"Leaderboards are not available for this beatmap!"));
+                        break;
 
-                    PlaceholderState = PlaceholderState.NetworkFailure;
-                });
+                    case PlaceholderState.NoScores:
+                        replacePlaceholder(new MessagePlaceholder(@"No records yet!"));
+                        break;
 
-                api?.Queue(getScoresRequest);
-            });
+                    case PlaceholderState.NotLoggedIn:
+                        replacePlaceholder(new LoginPlaceholder(@"Please sign in to view online leaderboards!"));
+                        break;
+
+                    case PlaceholderState.NotSupporter:
+                        replacePlaceholder(new MessagePlaceholder(@"Please invest in an osu!supporter tag to view this leaderboard!"));
+                        break;
+
+                    default:
+                        replacePlaceholder(null);
+                        break;
+                }
+            }
         }
 
-        /// <summary>
-        /// Performs a fetch/refresh of scores to be displayed.
-        /// </summary>
-        /// <param name="scoresCallback">A callback which should be called when fetching is completed. Scheduling is not required.</param>
-        /// <returns>An <see cref="APIRequest"/> responsible for the fetch operation. This will be queued and performed automatically.</returns>
-        protected abstract APIRequest FetchScores(Action<IEnumerable<TScoreInfo>> scoresCallback);
+        private void updateScoresDrawables()
+        {
+            currentScoresAsyncLoadCancellationSource?.Cancel();
 
-        private Placeholder currentPlaceholder;
+            scoreFlowContainer?
+                .FadeOut(fade_duration, Easing.OutQuint)
+                .Expire();
+            scoreFlowContainer = null;
+
+            loading.Hide();
+
+            if (scores?.Any() != true)
+            {
+                PlaceholderState = PlaceholderState.NoScores;
+                return;
+            }
+
+            // ensure placeholder is hidden when displaying scores
+            PlaceholderState = PlaceholderState.Successful;
+
+            LoadComponentAsync(new FillFlowContainer<LeaderboardScore>
+            {
+                RelativeSizeAxes = Axes.X,
+                AutoSizeAxes = Axes.Y,
+                Spacing = new Vector2(0f, 5f),
+                Padding = new MarginPadding { Top = 10, Bottom = 5 },
+                ChildrenEnumerable = scores.Select((s, index) => CreateDrawableScore(s, index + 1))
+            }, newFlow =>
+            {
+                scrollContainer.Add(scoreFlowContainer = newFlow);
+
+                int i = 0;
+
+                foreach (var s in scoreFlowContainer.Children)
+                {
+                    using (s.BeginDelayedSequence(i++ * 50))
```

Comment:
Would you prefer `i += 50` in this case?

Reply:
It wouldn't be the same thing, would it, though? To keep this exactly as is you'd have to do

```csharp
using (s.BeginDelayedSequence(i * 50)
    s.Show();
i += 1;
```

which shows the problem with the original code (it tries to be a bit too clever for my liking).

Reply:
Sure I'll convert it to plain math. We should probably discuss this at some point since you take issue with this but I prefer the way it's written there. I'll try and stop doing that for this project to keep everyone happy.

Reply:
Should probably tag in @smoogipoo for this. Maybe it's just me finding this difficult, in which case I'll bite the bullet and attempt to tolerate these.

Reply:
Definitely like the way it is currently in this PR (`delay += 50` after iteration) than in the above comment or in the original diff.